docs/DevGuide/Brigand.md

```diff
@@ -0,0 +1,1823 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::real_
+
+\par
+There are no compile-time mathematical functions provided for floating point
+types.  They are opaque (or sometimes treated as integers) until runtime.
+
+\remark
+Consider whether you really need to represent floating point values at compile
+time.
+
+\see std::ratio
+
+
+\subsubsection repeat
+
+\par
+Calls a unary eager metafunction on a given type, then on the result of that,
+then on the result of that, and so on, up to a specified (as an \ref
+integral_constant) number of calls.
+\snippet Test_TMPLDocumentation.cpp tmpl::repeat
+
+\note
+This function has a lazy version, but it cannot be used in a metalambda because
+the template template parameter prevents manipulation of the parameter list.
+\snippet Test_TMPLDocumentation.cpp tmpl::repeat:lazy
+
+\see \ref make_sequence
+
+
+\subsubsection sizeof_
+
+\par
+A lazy metafunction that computes `sizeof` its argument as an `unsigned int`
+\ref integral_constant.
+\snippet Test_TMPLDocumentation.cpp tmpl::sizeof_
+
+
+\subsubsection substitute
+
+\par
+Substitutes values for \ref args "args" (but *not* `_1` or `_2`) appearing in a
+type.
+\snippet Test_TMPLDocumentation.cpp tmpl::substitute
+
+
+\subsubsection type_from
+
+\par
+Extracts the `type` from a struct.
+\snippet Test_TMPLDocumentation.cpp tmpl::type_from
+
+\remark
+This function will work on any class with a `type` type alias, but, when used
+outside of a metafunction, it should only be used with \ref type_ for clarity.
+
+
+\subsubsection wrap
+
+\par
+Replaces the head of a sequence.
+\snippet Test_TMPLDocumentation.cpp tmpl::wrap
+
+\note
+This function has a lazy version, but it cannot be used in a metalambda because
+the template template parameter prevents manipulation of the parameter list.
+\snippet Test_TMPLDocumentation.cpp tmpl::wrap:lazy
+
+
+\subsection runtime Runtime functionality
+
+\par
+Brigand provides a few C++ functions that execute at runtime.
+
+\par
+The examples in this section use the following definition:
+\snippet Test_TMPLDocumentation.cpp runtime_declarations
+
+
+\subsubsection for_each_args
+
+\par
+Calls the first argument (a functor) on each of the remaining arguments, in
+order.  Returns the functor.
+\snippet Test_TMPLDocumentation.cpp tmpl::for_each_args:defs
+\snippet Test_TMPLDocumentation.cpp tmpl::for_each_args
+
+\note
+The functor must be copyable.  This is a bug.
+
+\par
+
+\note
+This uses a std::reference_wrapper internally, but I don't see a reason for
+that.  If it were removed then this function could be constexpr (before C++20).
+
+
+\subsubsection for_each
+
+\par
+Calls a functor on \ref type_ objects wrapping each type in a list, in order.
+Returns the functor.
+\snippet Test_TMPLDocumentation.cpp tmpl::for_each:defs
+\snippet Test_TMPLDocumentation.cpp tmpl::for_each
+
+\note
+The functor must be copyable.  This is a bug.
+
+\par
+
+\note
+An object of the list template parameter type is constructed, so the list must
```

Comment:
Does this mean the elements of the list must all be complete types? Looking at the Brigand implementation, I think if instead `for_each_impl` took a `List<Elements...>*` would that fix the need for having complete types? Otherwise, I think using a `struct` partial specialization could also fix the issue, right?

This is mostly for my understanding :)